compiler/test/dotty/tools/dotc/ParallelTesting.scala

```diff
@@ -0,0 +1,1076 @@
+          meth.invoke(null, Array("jvm")) // partest passes at least "jvm" as an arg
+        }
+
+        val outputLines = printStream.toString("utf-8").lines.toArray
+        val checkLines = Source.fromFile(checkFile).getLines.toArray
+
+        def linesMatch =
+          outputLines
+          .zip(checkLines)
+          .forall { case (x, y) => x == y }
+
+        if (outputLines.length != checkLines.length || !linesMatch) {
+          // Print diff to files and summary:
+          val diff = outputLines.zip(checkLines).map { case (act, exp) =>
+            DiffUtil.mkColoredCodeDiff(exp, act, true)
+          }.mkString("\n")
+          val msg = s"\nOutput from run test '$checkFile' did not match expected, output:\n$diff\n"
+          echo(msg)
+          addFailureInstruction(msg)
+
+          // Print build instructions to file and summary:
+          val buildInstr = target.buildInstructions(0, warnings)
+          addFailureInstruction(buildInstr)
+
+          // Fail target:
+          failTarget(target)
+        }
+    }
+    catch {
+      case _: NoSuchMethodException =>
+        echo(s"\ntest in '$dir' did not contain a main method")
+        fail()
+
+      case _: ClassNotFoundException =>
+        echo(s"\ntest in '$dir' did was not contained within a `Test` object")
+        fail()
+
+      case _: InvocationTargetException =>
+        echo(s"\nTest in '$dir' might be using args(X) where X > 0")
+        fail()
+    }
+
+    protected def compilationRunnable(target: Target): Runnable = new Runnable {
+      def run(): Unit = compileTry {
+        val (errorCount, warningCount, hasCheckFile, doVerify) = target match {
+          case ConcurrentCompilationTarget(files, flags, outDir) => {
+            val sourceFiles = files.filter(isCompilable)
+            val checkFile = files.flatMap { file =>
+              if (file.isDirectory) Nil
+              else {
+                val fname = file.getAbsolutePath.reverse.dropWhile(_ != '.').reverse + "check"
+                val checkFile = new JFile(fname)
+                if (checkFile.exists) List(checkFile)
+                else Nil
+              }
+            }.headOption
+            val reporter = compile(sourceFiles, flags, false, outDir)
+
+            if (reporter.errorCount > 0) {
+              fail()
+              val errorMsg = target.buildInstructions(reporter.errorCount, reporter.warningCount)
+              addFailureInstruction(errorMsg)
+              failTarget(target)
+              reporter.echo(errorMsg)
+              reporter.flushToFile()
+            }
+
+            completeCompilation(reporter.errorCount)
+            (reporter.errorCount, reporter.warningCount, checkFile.isDefined, () => verifyOutput(checkFile.get, outDir, target, reporter.warningCount))
+          }
+
+          case target @ SeparateCompilationTarget(dir, flags, outDir) => {
+            val checkFile = new JFile(dir.getAbsolutePath.reverse.dropWhile(_ == '/').reverse + ".check")
```

Comment:
For separate compilation folders we can have ".check" files, but we can't have them for joint compilation. Why?